dashboard/client/src/components/DataOverviewTable.tsx

```diff
@@ -116,59 +123,136 @@ const DataOverviewTable = ({
-                state,
-                ray_data_current_bytes,
-                ray_data_output_bytes,
-                ray_data_spilled_bytes,
-                progress,
-                total,
-                start_time,
-                end_time,
-              }) => (
-                <TableRow key={dataset}>
-                  <TableCell align="center">{dataset}</TableCell>
-                  <TableCell align="center">
-                    <TaskProgressBar
-                      numFinished={progress}
-                      numRunning={
-                        state === "RUNNING" ? total - progress : undefined
-                      }
-                      numCancelled={
-                        state === "FAILED" ? total - progress : undefined
-                      }
-                      total={total}
-                    />
-                  </TableCell>
-                  <TableCell align="center">
-                    <StatusChip type="task" status={state} />
-                  </TableCell>
-                  <TableCell align="center">
-                    {memoryConverter(Number(ray_data_output_bytes.max))}
-                  </TableCell>
-                  <TableCell align="center">
-                    {memoryConverter(Number(ray_data_current_bytes.value))}/
-                    {memoryConverter(Number(ray_data_current_bytes.max))}
-                  </TableCell>
-                  <TableCell align="center">
-                    {memoryConverter(Number(ray_data_spilled_bytes.max))}
-                  </TableCell>
-                  <TableCell align="center">
-                    {formatDateFromTimeMs(start_time * 1000)}
-                  </TableCell>
-                  <TableCell align="center">
-                    {end_time && formatDateFromTimeMs(end_time * 1000)}
-                  </TableCell>
-                </TableRow>
-              ),
-            )}
+            {list.map((dataset, index) => (
+              <DatasetTable
+                datasetMetrics={dataset}
+                isExpanded={expandedDatasets[dataset.dataset]}
+                setIsExpanded={(isExpanded: boolean) => {
+                  const copy = {
+                    ...expandedDatasets,
+                    [dataset.dataset]: isExpanded,
+                  };
+                  setExpandedDatasets(copy);
+                }}
+                key={dataset.dataset}
+              />
+            ))}
           </TableBody>
         </Table>
       </div>
     </div>
   );
 };
 
+const useStyles = makeStyles((theme) =>
+  createStyles({
+    icon: {
+      width: 16,
+      height: 16,
+    },
+  }),
+);
+
+const DataRow = ({
+  data,
+  isDatasetRow,
+  isExpanded,
+  setIsExpanded,
+}: {
+  data: DatasetMetrics | OperatorMetrics;
+  isDatasetRow: boolean;
+  isExpanded?: boolean;
+  setIsExpanded?: CallableFunction;
+}) => {
+  const classes = useStyles();
+
+  const dataset = data as DatasetMetrics;
+  const operator = data as OperatorMetrics;
```

Comment:
there's a few ways you can avoid casting these values.

One way is you can make the props like this:
```
({ data: DatasetMetrics, isDatasetRow: true } | { data: OperatorMetrics, isDatasetRow: false }) & { setExpanded?: boolean, setIsExpanded?: CallableFunction; }
```

Another way is you can create a [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) to check if data is DatasetMetrics or OperatorMetrics

A third way is you can simply have your props be
```
{
  datasetMetrics?: DatasetMetrics;
  operatorMetrics?: OperaterMetrics;
  isDatasetRow: boolean;
  isExpanded?: boolean;
  setIsExpanded?: CallableFunction;
}
```
and assert if not at least one is passed in.

All of these are preferred over using "as" since as will skip type checking completely.

Reply:
thanks for the alternatives!